### ISSUE TITLE
Sitecore hangs after SendSerializationCompleteEvent is triggered

### ISSUE THREAD

**Issue by erikvanbeijnen** [SitecoreUnicorn/Unicorn#378]
We are upgrading to Sitecore 9.3 and using latest Unicorn (4.1.1).
After running a Unicorn sync the Sitecore CMS hangs. Only solution is to perform an ap-pool recycle.

Investigation turns out that this issue is triggered by the SendSerializationCompleteEvent. When removing that from the pipeine, Unicorn sync works without hanging Sitecore.

We see the following in the logs:

> 25088 10:00:40 INFO  Job started: Raise deserialization complete async (Unicorn.Pipelines.UnicornSyncEnd.SendSerializationCompleteEvent)
25088 10:00:40 INFO  Job started: Raise deserialization complete async (Unicorn.Pipelines.UnicornSyncEnd.SendSerializationCompleteEvent)
...
27324 10:00:40 ERROR Cannot find property with 'QueueEvent' method. SerializationFinishedEvent is not rised.
29672 10:00:40 ERROR Cannot find property with 'QueueEvent' method. SerializationFinishedEvent is not rised.

Not sure if this is a bug or just a misconfiguration?

**Comment by cassidydotdk** [comment]
Well it's not misconfiguration ;-)

I've recently upgraded my Unicorn test instance to 9.3, I've been hearing various small reports about potential issues. Will look into this tomorrow (I hope) and let you know.

**Comment by cassidydotdk** [comment]
I can confirm part of this. I don't experience Sitecore hangs.

```
3908 11:11:42 INFO  [Rainbow] SFS tree item C:\inetpub\wwwroot\App_Data\Unicorn\Test\Content\content\Home.yml changed (ChangeOrAdd), caches updating.
3724 11:11:45 INFO  [Unicorn]: Test is being synced.
3724 11:11:45 INFO  Cache created: 'ItemCloningRelations' (max size: 50MB, running total: 2732MB)
3724 11:11:45 INFO  [Unicorn]: * [U] en: __Display name
3724 11:11:45 INFO  [Unicorn]: * [U] en#1: Text
3724 11:11:45 INFO  [Unicorn]: [U] master:/sitecore/content/Home (110d559f-dea5-42ea-9c1c-8a5df7e70ef9)
3724 11:11:45 INFO  [Unicorn]: > Field __Display name - en - Serialized 'Test Display Name', Source ''
3724 11:11:45 INFO  [Unicorn]: Test sync complete: 4 items evaluated, 1 item modified (0 added, 1 updated, 0 recycled) in 147ms (~36.94ms/item).
3724 11:11:45 INFO  [Unicorn]: 
3724 11:11:45 WARN  All caches have been cleared. This can decrease performance considerably. DEBUG entries have to be enabled in log4net to see the corresponding stack trace.
3724 11:11:45 INFO  Unicorn: initiated synchronous publishing of synced items.
3724 11:11:45 INFO  Cache created: 'SqlDataProvider - Prefetch data(web)' (max size: 100MB, running total: 2832MB)
3724 11:11:46 INFO  SmartHtmlCacheClearer: HTML cache has been cleared for 2 sites.
3724 11:11:46 INFO  RenderingParametersCacheClearer clearing caches for the sites: website.
3724 11:11:46 INFO  RenderingParametersCacheClearer done.
3724 11:11:46 INFO  Unicorn: publishing of synced items is complete.
3724 11:11:46 INFO  Job started: Raise deserialization complete async (Unicorn.Pipelines.UnicornSyncEnd.SendSerializationCompleteEvent)
3032 11:11:46 WARN  All caches have been cleared. This can decrease performance considerably.
3032 11:11:46 ERROR Cannot find property with 'QueueEvent' method. SerializationFinishedEvent is not rised.
```

**Comment by cassidydotdk** [comment]
Fixed.

```
3276 12:25:50 INFO  [Rainbow] SFS tree item C:\inetpub\wwwroot\App_Data\Unicorn\Test\Content\content\Home.yml changed (ChangeOrAdd), caches updating.
ManagedPoolThread #3 12:25:51 INFO  Job started: Sitecore.ContentSearch.SolrProvider.Agents.IndexingStateSwitcher
ManagedPoolThread #3 12:25:51 INFO  Job ended: Sitecore.ContentSearch.SolrProvider.Agents.IndexingStateSwitcher (units processed: )
 700 12:25:56 INFO  [Unicorn]: Test is being synced.
 700 12:25:56 INFO  [Unicorn]: * [U] en: __Display name
 700 12:25:56 INFO  [Unicorn]: [U] master:/sitecore/content/Home (110d559f-dea5-42ea-9c1c-8a5df7e70ef9)
 700 12:25:56 INFO  [Unicorn]: > Field __Display name - en - Serialized 'Test Display Name Updated', Source 'Test Display Name'
 700 12:25:56 INFO  [Unicorn]: Test sync complete: 4 items evaluated, 1 item modified (0 added, 1 updated, 0 recycled) in 108ms (~27.18ms/item).
 700 12:25:56 INFO  [Unicorn]: 
 700 12:25:56 WARN  All caches have been cleared. This can decrease performance considerably. DEBUG entries have to be enabled in log4net to see the corresponding stack trace.
 700 12:25:56 INFO  Unicorn: initiated synchronous publishing of synced items.
 700 12:25:56 INFO  Cache created: 'SqlDataProvider - Prefetch data(web)' (max size: 100MB, running total: 2832MB)
 700 12:25:57 INFO  SmartHtmlCacheClearer: HTML cache has been cleared for 2 sites.
 700 12:25:57 INFO  RenderingParametersCacheClearer clearing caches for the sites: website.
 700 12:25:57 INFO  RenderingParametersCacheClearer done.
 700 12:25:57 INFO  Unicorn: publishing of synced items is complete.
 700 12:25:57 INFO  Job started: Raise deserialization complete async (Unicorn.Pipelines.UnicornSyncEnd.SendSerializationCompleteEvent)
 700 12:25:57 WARN  All caches have been cleared. This can decrease performance considerably.
 700 12:25:57 ERROR Cannot find property with 'QueueEvent' method. SerializationFinishedEvent is not raised.
```

**Comment by cassidydotdk** [comment]
Fixed. Will publish a release shortly.

```
3428 12:38:29 INFO  [Rainbow] SFS tree item C:\inetpub\wwwroot\App_Data\Unicorn\Test\Content\content\Home.yml changed (ChangeOrAdd), caches updating.
3696 12:38:30 INFO  [Unicorn]: Test is being synced.
3696 12:38:30 INFO  Cache created: 'ItemCloningRelations' (max size: 50MB, running total: 2482MB)
3696 12:38:30 INFO  [Unicorn]: * [U] en: __Display name
3696 12:38:30 INFO  [Unicorn]: [U] master:/sitecore/content/Home (110d559f-dea5-42ea-9c1c-8a5df7e70ef9)
3696 12:38:30 INFO  [Unicorn]: > Field __Display name - en - Serialized 'And again!', Source 'And again'
3696 12:38:30 INFO  [Unicorn]: Test sync complete: 4 items evaluated, 1 item modified (0 added, 1 updated, 0 recycled) in 227ms (~56.98ms/item).
3696 12:38:30 INFO  [Unicorn]: 
3696 12:38:30 WARN  All caches have been cleared. This can decrease performance considerably. DEBUG entries have to be enabled in log4net to see the corresponding stack trace.
3696 12:38:30 INFO  Unicorn: initiated synchronous publishing of synced items.
3696 12:38:30 INFO  Cache created: 'SqlDataProvider - Prefetch data(web)' (max size: 100MB, running total: 2582MB)
3696 12:38:31 INFO  SmartHtmlCacheClearer: HTML cache has been cleared for 2 sites.
3696 12:38:31 INFO  RenderingParametersCacheClearer clearing caches for the sites: website.
3696 12:38:31 INFO  RenderingParametersCacheClearer done.
3696 12:38:31 INFO  Unicorn: publishing of synced items is complete.
3696 12:38:31 INFO  Job started: Raise deserialization complete async (Unicorn.Pipelines.UnicornSyncEnd.SendSerializationCompleteEvent)
3696 12:38:31 WARN  All caches have been cleared. This can decrease performance considerably.
3696 12:38:31 INFO  Job ended: Raise deserialization complete async (Unicorn.Pipelines.UnicornSyncEnd.SendSerializationCompleteEvent)
```

**Comment by cassidydotdk** [comment]
Fixed in https://github.com/SitecoreUnicorn/Unicorn/commit/3f37b49fa01f4b82bad3cc747ba712a2bed34d9c

**Comment by cassidydotdk** [comment]
This has been released https://github.com/SitecoreUnicorn/Unicorn/releases/tag/4.1.2-pre1

It should appear on the NuGet Gallery shortly. 

Please let me know if this solves the issue for you.

**Comment by erikvanbeijnen** [comment]
Thanks for the quick fix!
This pre-release fixes the error, but unfortunately Sitecore still hangs after a sync :(
Probably this is caused by misconfiguration in our solution, not a Unicorn-related issue. I will close this issue.

**Comment by paulgeorge** [comment]
> Thanks for the quick fix!
> This pre-release fixes the error, but unfortunately Sitecore still hangs after a sync :(
> Probably this is caused by misconfiguration in our solution, not a Unicorn-related issue. I will close this issue.

@erikvanbeijnen did you ever resolve the hang after sync issue? I see the same behaviour with the same message.

**Comment by erikvanbeijnen** [comment]
Hi @paulgeorge, since it was a year ago I experienced this issue I needed to look it up.
Eventually we were able to pinpoint this to Sitecore specific code (not Unicorn related).
There seems to be a Sitecore bug in version 9.3 that after running SerializationFinishedEvent too many caches would be cleared. 

In a small Siteocore sulution this just results in slight a longer startup time after running this event, but since we have a large solution and database our solution just seems to 'hang'.
Sitecore Support eventually delivered us a hotfix which resolved this issue for us.

Maybe you are facing the same bug. I would like to advice you to create a Sitecore support ticket. The Sitecore bug reference number for this bug is 96704.
